proof/refine/ARM/Reply_R.thy

```diff
@@ -96,4 +95,487 @@ lemma replyRemoveTCB_st_tcb_at'_sym_ref:
+                         projectKO_eq
+                  split: option.split)+
+  done
+
+lemma cleanReply_list_refs_of_replies':
+  "\<lbrace>\<lambda>s. P ((list_refs_of_replies' s)(rptr := {}))\<rbrace>
+   cleanReply rptr
+   \<lbrace>\<lambda>_ s. P (list_refs_of_replies' s)\<rbrace>"
+  unfolding cleanReply_def updateReply_def
+  apply (wpsimp wp: setReply_list_refs_of_replies')
+  apply (erule rsubst[where P=P])
+  apply (rule ext)
+  apply (clarsimp simp: list_refs_of_reply'_def fun_upd_def)
+  done
+
+lemma isHead_to_head:
+  "isHead x = (\<exists>scPtr. x = Some (Head scPtr))"
+  "(\<forall>scPtr. y \<noteq> Head scPtr) = (\<exists>rPtr. y = Next rPtr)"
+  apply (clarsimp simp: isHead_def split: option.split reply_next.split)
+  apply (case_tac y; clarsimp)
+  done
+
+(* FIXME RT: remove. proved by MikiT *)
+lemma sym_refs_replyNext_replyPrev_proj:
+  "sym_refs (list_refs_of_replies' s') \<Longrightarrow>
+    replyNexts_of s' rp = Some rp' \<longleftrightarrow> replyPrevs_of s' rp' = Some rp"
+  sorry
+
+(* FIXME RT: remove. proved by MikiT *)
+lemma sym_refs_replyNext_None:
+  "\<lbrakk>sym_refs (list_refs_of_replies' s');
+    replyNexts_of s' rp = None \<rbrakk> \<Longrightarrow> \<forall>rp'. replyPrevs_of s' rp' \<noteq> Some rp"
+  sorry
+
+(* FIXME RT: remove. proved by MikiT *)
+lemma sym_refs_replyPrev_None:
+  "\<lbrakk>sym_refs (list_refs_of_replies' s');
+    replyPrevs_of s' rp = None \<rbrakk> \<Longrightarrow> \<forall>rp'. replyNexts_of s' rp' \<noteq> Some rp"
+  sorry
+
+lemma ks_reply_at'_repliesD:
+  "\<lbrakk>ks_reply_at' reply rptr s; sym_refs (list_refs_of_replies' s)\<rbrakk>
+   \<Longrightarrow> replyNexts_of s rptr = replyNext_of reply
+       \<and> (case replyNext_of reply of
+            Some next \<Rightarrow> replyPrevs_of s next = Some rptr
+          | None \<Rightarrow> \<forall>rptr'. replyPrevs_of s rptr' \<noteq> Some rptr)
+       \<and> replyPrevs_of s rptr = replyPrev reply
+       \<and> (case replyPrev reply of
+            Some prev \<Rightarrow> replyNexts_of s prev = Some rptr
+          | None \<Rightarrow> \<forall>rptr'. replyNexts_of s rptr' \<noteq> Some rptr)"
+  \<comment>\<open> opt_mapE will sometimes destroy the @{term "(|>)"} inside @{term replyNexts_of}
+      and @{term replyPrevs_of}, but we're using those as our local normal form when proving
+      facts about @{term "sym_refs o list_refs_of_replies'"}. \<close>
+  supply opt_mapE[rule del]
+  apply (prop_tac "replyNexts_of s rptr = replyNext_of reply
+                   \<and> replyPrevs_of s rptr = replyPrev reply ")
+   apply (clarsimp simp: opt_map_def reply_of'_KOReply)
+  apply (case_tac "replyNext_of reply"
+         ; case_tac "replyPrev reply"
+         ; clarsimp simp: sym_refs_replyNext_None sym_refs_replyPrev_None
+                          sym_refs_replyNext_replyPrev_proj)
+  done
+
+\<comment>\<open> Used to "hide" @{term "sym_refs o list_refs_of_replies'"} from simplification. \<close>
```

Comment:
Are there not more standard ways of doing the same sort of thing? `thin_tac` maybe? I guess if the use of this is limited it's not so bad

Reply:
`thin_tac` doesn't have a way to only temporarily hide assumptions: see [VER-1244](https://sel4.atlassian.net/browse/VER-1244), which is exactly about this. Sadly, dodgy temporary definitions seem to be the easiest way to do this sort of thing.

Reply:
Get workin' on that JIRA issue! Sounds good